спринт 16.

pyclan88 opened the pull request:
проект полностью переделан под требования MVVM и Clean Architecture

pyclan88 committed:
спринт 16.
проект полностью переделан под требования MVVM и Clean Architecture

kekulta commented on app/src/main/java/com/practicum/playlistmaker/player/ui/view_model/PlayerViewModel.kt:
```diff
@@ -0,0 +1,94 @@
+    private val playerInteractor: PlayerInteractor
+) : ViewModel() {
+
+    companion object {
+        private const val DELAY_MILLIS = 300L
+        private val PLAYER_REQUEST_TOKEN = Any()
+
+        fun getViewModelFactory(context: Context): ViewModelProvider.Factory = viewModelFactory {
+            initializer {
+                PlayerViewModel(
+                    Creator.providePlayerInteractor(context)
+                )
+            }
+        }
+    }
```
Согласно [код-стайлу](https://kotlinlang.org/docs/coding-conventions.html#class-layout) Котлина Companion Object должен находится в конце класса.
Хотя я лично знаю команды, где он пишется в начале класса.
В целом, код-стайл это не про жёсткие правила, а про договорённости внутри
команды. Но я бы рекомендовал использовать стандартные практики в новых
проектах, где ещё не устоялись нестадартные. Это будет очень полезно при
онбординге новых разработчиков.

В целом рекомендую внимательно ознакомиться с Coding Conventions Котлина и по
возможности им следовать.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/search/data/impl/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,41 @@
+        val response = networkClient.doRequest(SearchRequest(expression))
+        return when (response.resultCode) {
+            -1 -> {
+                Resource.Error("Проблемы со связью\nЗагрузка не удалась. Проверьте\n" +
+                        "подключение к интернету")
```
Нужно вынести в строковые ресурсы.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/util/Resource.kt:
```diff
@@ -0,0 +1,6 @@
+package com.practicum.playlistmaker.util
+
+sealed class Resource<T>(val data: T? = null, val message: String? = null) {
+    class Success<T>(data: T) : Resource<T>(data)
+    class Error<T>(message: String, data: T? = null) : Resource<T>(data, message)
+}
```
Очень здорово, что используешь Resource класс! Подобные "монады" могут быть очень
полезными. Рекомендую ознакомиться с библиотекой [kotlin-result](https://github.com/michaelbull/kotlin-result).
Она может быть очень полезной в будущем. Так же имплементация не сложная,
можно в ней разобраться и подчерпнуть для себя полезных идей.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/search/ui/activity/SearchActivity.kt:
```diff
@@ -0,0 +1,183 @@
+        }
+    }
+
+    private fun showLoading() {
+        binding.progressBar.isVisible = true
+        binding.youSearched.isVisible = false
+        binding.clearHistory.isVisible = false
+        binding.searchScroll.isVisible = false
+        binding.linearNothingFound.isVisible = false
+        binding.linearInternetError.isVisible = false
+    }
```
Можно написать экстеншены `View.gone()`, `View.visible()`, etc.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/search/domain/impl/TrackDbInteractorImpl.kt:
```diff
@@ -0,0 +1,24 @@
+import com.practicum.playlistmaker.search.domain.api.TrackDbRepository
+import com.practicum.playlistmaker.search.domain.TrackDbInteractor
+
+class TrackDbInteractorImpl(private val trackDBRepository: TrackDbRepository) : TrackDbInteractor {
```
Не очень хорошая идея выносить DB в название интерактора. Интерактор должен быть агностичен относительно методов хранения, это детали реализации.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/search/ui/view_model/SearchViewModel.kt:
```diff
@@ -0,0 +1,110 @@
+package com.practicum.playlistmaker.search.ui.view_model
```
Не очень хорошая практика использовать нижнее подчёркивание в  названиях пакетов.

kekulta commented on app/src/main/java/com/practicum/playlistmaker/search/ui/view_model/SearchViewModel.kt:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    private fun renderState(state: SearchScreenState) {
+        screenStateLiveData.postValue(state)
+    }
```
render это скорее во View. Здесь будет set или send state.

pyclan88 committed:
спринт 16. исправлены недочеты после ревью

pyclan88 committed:
спринт 16
- исправлено отсутствие истории при нажатии на крестик или если стереть запрос с помощью клавиатуры
- исправлены недочеты

pyclan88 committed:
спринт 16
- исправлено отсутствие истории при нажатии на крестик или если стереть запрос с помощью клавиатуры
- исправлены недочеты

pyclan88 committed:
спринт 16
- исправлено отсутствие истории при нажатии на крестик или если стереть запрос с помощью клавиатуры
- исправлены недочеты
